src/main/java/bio/terra/service/filedata/flight/ingest/IngestPopulateFileStateFromFileStep.java

```diff
@@ -51,26 +60,40 @@ void readFile(BufferedReader reader, FlightContext context) throws IOException {
         new ErrorCollector(
             maxBadLoadFileLineErrorsReported,
             "Invalid lines in the control file. [All lines in control file must be valid in order to proceed - 'maxFailedFileLoads' not applicable here.]");
-    long lineCount = 0;
-    List<BulkLoadFileModel> fileList = new ArrayList<>();
+    List<Future<BulkLoadFileModel>> futures = new ArrayList<>();
 
+    // Value used in a lambda so needs to be effectively final
+    final AtomicLong lineCount = new AtomicLong(0);
     String line;
     while ((line = reader.readLine()) != null) {
-      lineCount++;
+      final String lineCopy = line;
+      lineCount.incrementAndGet();
 
-      try {
-        BulkLoadFileModel loadFile = bulkLoadObjectMapper.readValue(line, BulkLoadFileModel.class);
-        IngestUtils.validateBulkLoadFileModel(loadFile);
-        cloudFileReader.validateUserCanRead(List.of(loadFile.getSourcePath()), userRequest);
-        fileList.add(loadFile);
-      } catch (IOException | BlobAccessNotAuthorizedException | BadRequestException ex) {
-        errorCollector.record("Error at line %d: %s", lineCount, ex.getMessage());
-      }
+      // Run batches in parallel
+      futures.add(
+          executor.submit(
+              () -> {
+                try {
+                  BulkLoadFileModel loadFile =
+                      bulkLoadObjectMapper.readValue(lineCopy, BulkLoadFileModel.class);
+                  IngestUtils.validateBulkLoadFileModel(loadFile);
+                  cloudFileReader.validateUserCanRead(
```

Comment:
I wonder if there's an opportunity to be smarter here. We check each individual file to see if we have permission to read it. If all the files are in the same bucket, and the user has whole-bucket read access, could we just check for that instead, and repeat once for each bucket encountered? It would be more complicated, but would mean not checking `n` times, where `n` is the number of files being ingested.

Reply:
I'm not totally sure that there is.  That being said, I think ultimately, there's still an upper bound here and we likely need to think of different auth methods

Reply:
> all the files are in the same bucket, and the user has whole-bucket read access, could we just check for that instead, and repeat once for each bucket encountered?

It looks like `GcsPdao.validateUserCanRead()` is already doing this. If this code was written to pass all the files at once it would take advantage of that.

Reply:
That's for the overall control file.  We also need to be careful since the file list can get really big

Reply:
We could pass batches of files to validateUserCanRead() - I think that could cut down on the time.

Reply:
Im not sure...I think that the one in GcsPdao isn't parallel...One thing I just noticed: this should actually get faster in prod given that the GCS check doesn't actually happen in connected tests.  I'm going to merge in to make sure it gets into the next release

Reply:
> That's for the overall control file. We also need to be careful since the file list can get really big

There's two different validate calls here which makes this discussion confusing. The load file check (`IngestUtils.validateBulkLoadFileModel`) doesn't accept multiple arguments but it could easily do so if wanted to batch up requests. But that's not what I was commenting on.

Regarding `cloudFileReader.validateUserCanRead()`, it seems to me that there's a huge benefit to passing a list of files to this method instead of calling it in multiple threads, assuming that the same bucket is used in different load files in this operation.

Reply:
Also `AzureBlobStorePdao.validateUserCanRead()` is a no-op so running that in multiple threads is wasteful.

If there's a worry about number of files and a need for parallelism, that could be done in `GcsPdao` since GCS is currently the only platform that needs it.

Reply:
I'm hesitant to push it down further down since the bit that is building up the loop (e.g. the line reader is in the class that I modified) it seems like then the concurrency is hidden